### PR TITLE
GCM AAD and encrypted byte limit checks in AEADCipherContext

### PR DESCRIPTION
**Pull request by reaperhulk**
Like #1996 but up in the AEADCipherContext.

### CODE REVIEW

File: src/cryptography/hazmat/primitives/ciphers/base.py [hunk 1]
```diff
@@ -171,6 +181,15 @@ def authenticate_additional_data(self, data):
             raise AlreadyFinalized("Context was already finalized.")
         if self._updated:
             raise AlreadyUpdated("Update has been called on this context.")
+
+        self._aad_bytes_processed += len(data)
+        if self._aad_bytes_processed > modes.GCM._MAX_AAD_BYTES:
```
Comment (alex) [inline src/cryptography/hazmat/primitives/ciphers/base.py]: Not `self._ctx._mode` or something?

Comment (reaperhulk) [inline src/cryptography/hazmat/primitives/ciphers/base.py]: That'd be a mistake :D fixing

**Comment by alex** [comment]
Please add a test that the counters start at 0, that `update('   ')` advances it by 3, and then doing the same does it again (feel free to change the values and the counts :P)

File: tests/hazmat/primitives/test_aes.py [hunk 2]
```diff
@@ -253,3 +253,43 @@ def test_gcm_ciphertext_with_no_aad(self, backend):
...
         assert encryptor.tag == tag
+
+    def test_gcm_ciphertext_limit(self, backend):
+        encryptor = base.Cipher(
+            algorithms.AES(b"\x00" * 16),
+            modes.GCM(b"\x01" * 16),
+            backend=backend
+        ).encryptor()
+        # 16 bytes less than the encryption limit
+        near_limit_bytes = (2 ** 39 - 256 - 128) // 8
```
Comment (alex) [inline tests/hazmat/primitives/test_aes.py]: Any reason not to use `modes.GCM._MAX_ENCRYPTED_BYTES - 16`?

**Comment by alex** [comment]
I don't see a test for overflowing the aad count.

**Comment by alex** [comment]
lookin good

**Pull request merged**